# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/webauthn/WebAuthnHandler.cpp
# Commit: e1dc56ae7277
# Full Hash: e1dc56ae7277b29a3ecbd5ec3cd8612a3697b57a
# Author: John M. Schanck <jschanck@mozilla.com>
# Date: 2025-06-06 08:52:03
# Description:
#   Bug 1969341 - clean up WebAuthn transaction state before resolving promises. r=djackson
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D251966
# ==============================================================================

diff -r 81f3e9043996 -r e1dc56ae7277 dom/webauthn/WebAuthnHandler.cpp
--- a/dom/webauthn/WebAuthnHandler.cpp	Fri Jun 06 02:05:23 2025 +0000
+++ b/dom/webauthn/WebAuthnHandler.cpp	Fri Jun 06 04:20:30 2025 +0000
@@ -959,9 +959,14 @@
       break;
   }
 
-  mTransaction.ref().mPromise->MaybeResolve(aCredential);
+  // Bug 1969341 - we need to reset the transaction before resolving the
+  // promise. This lets us handle the case where resolving the promise initiates
+  // a new WebAuthn request.
+  RefPtr<Promise> promise = mTransaction.ref().mPromise;
   mTransaction.reset();
   Unfollow();
+
+  promise->MaybeResolve(aCredential);
 }
 
 template <typename T>
@@ -977,9 +982,14 @@
       break;
   }
 
-  mTransaction.ref().mPromise->MaybeReject(aReason);
+  // Bug 1969341 - we need to reset the transaction before rejecting the
+  // promise. This lets us handle the case where rejecting the promise initiates
+  // a new WebAuthn request.
+  RefPtr<Promise> promise = mTransaction.ref().mPromise;
   mTransaction.reset();
   Unfollow();
+
+  promise->MaybeReject(aReason);
 }
 
 }  // namespace mozilla::dom